KAFKA-8820: kafka-reassign-partitions.sh should support the KIP-455 API

cmccabe opened the pull request:
Rewrite ReassignPartitionsCommand to use the KIP-455 API when possible, rather
than direct communication with ZooKeeper.  Direct ZK access is still supported,
but deprecated, as described in KIP-555.

As specified in KIP-455, the tool has several new flags.  --cancel stops
an assignment which is in progress.  --preserve-throttle causes the
--verify and --cancel commands to leave the throttles alone.
--additional allows users to execute another partition assignment even
if there is already one in progress.  Finally, --show displays all of
the current partition reassignments.

Reorganize the reassignment code and tests somewhat to rely more on unit
testing using the MockAdminClient and less on integration testing.  Each
integration test where we bring up a cluster seems to take about 5 seconds, so
it's good when we can get similar coverage from unit tests.  To enable this,
MockAdminClient now supports incrementalAlterConfigs, alterReplicaLogDirs,
describeReplicaLogDirs, and some other APIs.  MockAdminClient is also now
thread-safe, to match the real AdminClient implementation.

In DeleteTopicTest, use the KIP-455 API rather than invoking the reassignment
command.

cmccabe committed:
KAFKA-8820: kafka-reassign-partitions.sh should support the KIP-455 API

Rewrite ReassignPartitionsCommand to use the KIP-455 API when possible, rather
than direct communication with ZooKeeper.  Direct ZK access is still supported,
but deprecated, as described in KIP-455.

As specified in KIP-455, the tool has several new flags.  --cancel stops
an assignment which is in progress.  --preserve-throttle causes the
--verify and --cancel commands to leave the throttles alone.
--additional allows users to execute another partition assignment even
if there is already one in progress.  Finally, --show displays all of
the current partition reassignments.

Reorganize the reassignment code and tests somewhat to rely more on unit
testing using the MockAdminClient and less on integration testing.  Each
integration test where we bring up a cluster seems to take about 5 seconds, so
it's good when we can get similar coverage from unit tests.  To enable this,
MockAdminClient now supports incrementalAlterConfigs, alterReplicaLogDirs,
describeReplicaLogDirs, and some other APIs.  MockAdminClient is also now
thread-safe, to match the real AdminClient implementation.

In DeleteTopicTest, use the KIP-455 API rather than invoking the reassignment
command.

cmccabe commented:
ok to test

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
   @After
   override def tearDown(): Unit = {
-    TestUtils.shutdownServers(servers)
+    Utils.closeQuietly(cluster , "ReassignPartitionsTestCluster")
```
nit: space

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
+            case (part, index) => partMap.put(index, part.map(Integer.valueOf(_)).asJava)
+          }
+          new NewTopic(topicName, partMap)
+      }.toList.asJava)
```
Should we ensure the createTopics call completes without an error here?

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
-        Map(topicPartition -> reassignmentEntry(Seq(6, broker2, broker3))).asJava
-      ).all()
-      JTestUtils.assertFutureThrows(future, classOf[InvalidReplicaAssignmentException])
+  def testReplicaMoves(): Unit = {
```
nit: Should we include the fact that replicas are moving in between directories in the test name? This would make the test failure output clearer on first read

cmccabe replied:
I changed the name to `testRplicaDirectoryMoves`

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
-    }
+import scala.collection.{Seq, mutable}
+
+class ReassignPartitionsIntegrationTest extends ZooKeeperTestHarness {
```
Should we leave one integration test that still uses the ZK API?

cmccabe replied:
added

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
+        new TopicPartition("baz", 1),
+        new TopicPartition("baz", 2)),
+      info0.curLogDirs.keySet)
+    val curFoo1Dir = info0.curLogDirs.get(new TopicPartition("foo", 0)).getOrElse("")
```
nit: IntelliJ recommends `Replace with getOrElse(key, defaultValue)`

stanislavkozlovski commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -1,175 +1,485 @@
+          assertEquals(Seq(0, 3, 2),
+            partStates.get(new TopicPartition("foo", 0)).get.targetReplicas)
+          assertEquals(Seq(3, 2, 1),
+            partStates.get(new TopicPartition("baz", 2)).get.targetReplicas)
```
nit: IntelliJ recommends `Replace with .(key)`

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   * @param sources         The source brokers.
+   * @param destinations    The destination brokers.
+   */
+  sealed case class PartitionMove(val sources: mutable.Set[Int],
```
nit: `val` is redundant for case classes as their values are immutable by default

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   *
+   * @param currentReplicas The current replicas.
+   * @param targetReplicas  The target replicas.
+   * @param done            True if the reassignment is still progressing.
```
Should this be `False if...` ?

cmccabe replied:
Fixed

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   */
+  sealed trait ReplicaMoveState {
+    /**
+     * True if the move is still progressing.
```
Should this be `False if...` ?

cmccabe replied:
Fixed

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+  }
+
+  /**
+   * A replica move state where the source log directory is missng.
```
nit: typo `missng`

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+  /**
+   * The state of a replica log directory movement.
+   */
+  sealed trait ReplicaMoveState {
```
What do you think we call this `LogDirMoveState` or something similar? `ReplicaMove` can be confused with a normal replica reassignment

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   */
+  sealed case class MissingReplicaMoveState(val targetLogDir: String)
+      extends ReplicaMoveState {
+    override def done = false
```
Should this be true given that no move is in progress, or should we update the docstring for the `done` variable to mention that error states are false

cmccabe replied:
As a practical matter, we probably don't want to remove the throttles in `--verify` if the log directory move is screwed up (because of an offline log directory or similar).  I will update the docstring.

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+        props.putIfAbsent(AdminClientConfig.CLIENT_ID_CONFIG, "reassign-partitions-tool")
+        val adminClient = Admin.create(props)
+        toClose = Some(adminClient)
+        // TODO: add --show and --cancel
```
Are we going to add `--show` or will we keep using`--list`? In any case, this comment seems leftover

cmccabe replied:
yes, this is left over.  removed

cmccabe replied:
I went with --list since it's more consistent with our other commands

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   *                                The partition map is keyed on partition index and contains
+   *                                the movements for that partition.
+   */
+  def calculateMoveMap(currentReassignments: Map[TopicPa
```
nit: `.nonEmpty` ?

cmccabe replied:
hmm, I'm not sure I follow.  Where do you think we should add `.nonEmpty`?

stanislavkozlovski replied:
Ugh, I don't remember. GitHub's UI broke. This was likely in one fo the places where we do `!condition.isEmpty` - IntelliJ glows and suggests doing `.nonEmpty`. I don't feel strongly about this

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   *                                The partition map is keyed on partition index and contains
+   *                                the movements for that partition.
+   */
+  def calculateMoveMap(currentReassignments: Map[TopicPa
```
Do we allow cancellation only of partition reassignments whose `addingReplicas` isn't empty or whose `removingReplicas` isn't empty? If so, the latter check needs a `!`

What about the alternative - allow cancellation of whatever `listPartitionReassignments()` returns? The API should return reassignments that are in progress only anyway

cmccabe replied:
Hmm, github is showing this comment in a weird place-- on the `calculateMoveMap` function rather than in the `cancelAssignment` function as you probably intended.

Good find on the missing negation.  That was definitely not intended.  Fixed.

If users want to cancel arbitrary partition moves, they can create a JSON file and use `--cancel` on it, right?

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+   *                                The partition map is keyed on partition index and contains
+   *                                the movements for that partition.
+   */
+  def calculateMoveMap(currentReassignments: Map[TopicPa
```
Should we print the ones that were successfully cancelled?

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+        done = true
+      } else if (time.milliseconds() >= startTimeMs + timeoutMs) {
+        throw new TerseReassignmentFailureException(
+          "Timed out before replica log dir(s) could be reassigned:%n%s".format(
```
If we get a `ReplicaNotAvailableException` for one replica throughout the whole `timeoutMs` duration , we'd print out that all movements failed. Shall we print `pendingReplicas`?

cmccabe replied:
That's a good idea.  In general, we should just print out all the partitions or log directories that were affected for each operation.

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1521 @@
+                        interBrokerThrottle: Long): Unit = {
+    val (proposedParts, proposedReplicas) = parseExecuteAssignmentArgs(reassignmentJson)
+    if (!proposedReplicas.isEmpty) {
+      throw new AdminCommandFailedException("bootstrap-server needs to be provided when " +
```
Do we have a test to ensure that the old way to run replica log dir reassignments with ZK works? i.e `./kafka-reassign-partitions --zookeeper ZK --bootstrap-server server --reassignment-json-file file_with_dir_reassignments.json`
From reading the code, I think so - but always worth making sure

cmccabe replied:
If you supply both `--zookeeper` and `--bootstrap-server`, then --zookeeper is ignored.  So there's no need for separate tests for the combination.

mumrah replied:
It would be nice if we gave a warning that the `--zookeeper` flag was being ignore in this case

stanislavkozlovski reviewed (commented):
Thanks for this massive PR! I'm happy to see the tool being updated. I gave this a quick pass (around 30-40%)

I really like the idea of moving most integration tests to unit - we had quite the heavy set of integration tests for this feature. That being said, I'm wondering whether we should retain a bit more of the reassignment integration tests. Namely some of the correctness checks - `ReassignPartitionsClusterTest #testHwAfterPartitionReassignment`, `testProduceAndConsumeWithReassignmentInProgress`, 
`shouldTriggerReassignmentOnControllerStartup` (could actually do `shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup` as it's a superset), `shouldDeleteReassignmentZnodeAfterApiReassignmentForPartitionCompletes` and maybe `shouldOnlyThrottleMovingReplicas`

The KIP mentions adding a `--create-cancellation-plan` config - something that essentially lists all running reassignments and creates a plan with which we could cancel them. Should we add it as part of this PR?

The KIP also mentions extending kafka-topics to show adding/removing replicas, but we can defer that to a follow-up JIRA I think.

With KIP-455 we added a TimeoutMs field in the RPC but did not implement it (tracked in https://issues.apache.org/jira/browse/KAFKA-8873). Would it be a good idea to include support for this field in this initial release of the reassign command to have forwards-compatibility? I guess not since it seems like it'd cause more confusion than not but worth raising the discussion nevertheless.

cmccabe commented:
Hi @stanislavkozlovski, thanks for the careful review.

> The KIP mentions adding a --create-cancellation-plan config - something that essentially lists all running reassignments and creates a plan with which we could cancel them. Should we add it as part of this PR?

Sorry that this was confusing.  I've updated the KIP with slightly revised command line options.  I think the new options make more sense and are simpler.  I'll post a message to the mailing list in a bit to confirm and get feedback.  In the meantime, check out the new text here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-455%3A+Create+an+Administrative+API+for+Replica+Reassignment

> With KIP-455 we added a TimeoutMs field in the RPC but did not implement it (tracked in https://issues.apache.org/jira/browse/KAFKA-8873). Would it be a good idea to include support for this field in this initial release of the reassign command to have forwards-compatibility? I guess not since it seems like it'd cause more confusion than not but worth raising the discussion nevertheless.

I think this should be handled internally by AdminClient, based on the standard timeout values that are configured.  In general, we want to minimize the number of timeout knobs.  Of course, as you mention, it's a moot point since it's not implemented yet :)

> The KIP also mentions extending kafka-topics to show adding/removing replicas, but we can defer that to a follow-up JIRA I think.

Thanks for the reminder.  Yes, let's do this in a separate PR.  I filed https://issues.apache.org/jira/browse/KAFKA-9688 as a reminder.

> I really like the idea of moving most integration tests to unit - we had quite the heavy set of integration tests for this feature. That being said, I'm wondering whether we should retain a bit more of the reassignment integration tests...

Yes, this is a good point.  I'll take a look.

cmccabe commented:
Thanks for the comments, @stanislavkozlovski .  I posted some updates to the PR.

There are some comments I haven't addressed yet, but I will soon.

cmccabe commented:
`testHwAfterPartitionReassignment`: I have added this test back in.

`testProduceAndConsumeWithReassignmentInProgress`: also added.

I also added an integration test of using the `--zookeeper` flag to start a reassignment.

`shouldOnlyThrottleMovingReplicas`: I think this is better handled by the unit test.

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1531 @@
+    val bld = new mutable.ArrayBuffer[String]()
+    bld.append("Status of partition reassignment:")
+    states.keySet.toBuffer.sortWith(compareTopicPartitions).foreach {
+      case topicPartition => {
```
Is this `case` necessary? Since there's no type binding and no "null" matches, wouldn't it just match everything?

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1531 @@
-        generateAssignment(zkClient, opts)
-      else if (opts.options.has(opts.executeOpt))
-        executeAssignment(zkClient, adminClientOpt, opts)
+      if (opts.options.has(opts.bootstrapServerOpt)) {
```
Maybe we could split this into two "main" methods: one for AdminClient and one for ZK? Might make this a bit more readable.

cmccabe replied:
ok

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1531 @@
+                         : (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
+    val (brokersToReassign, topicsToReassign) =
+      parseGenerateAssignmentArgs(reassignmentJson, brokerListString)
+    val currentAssignments = getReplicaAssignmentForTopics(adminClient, topicsToReassign)
```
Since these two `generateAssignment` methods are nearly identical, do you think it's worth trying to factor out the common parts? Maybe we could have a single `generateAssignment` that accepts an `Either[Admin, ZkClient]`?

cmccabe replied:
Good question.  The zookeeper code is going away soon, so I don't think it's worth trying to combine the functions.

I could write a function to factor out the "print" statements, but I'm not sure it's worth it since the main work is done by `calculateAssignment` and `formatAsReassignmentJson`, which are already shared code.

mumrah replied:
Yea, probably not worth the effort in this case.

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1531 @@
+    val (brokersToReassign, topicsToReassign) =
+      parseGenerateAssignmentArgs(reassignmentJson, brokerListString)
+    val currentAssignments = zkClient.getReplicaAssignmentForTopics(topicsToReassign.toSet)
+    val brokerMetadatas = if (enableRackAwareness) {
```
In the [old code](https://github.com/apache/kafka/pull/8244/files#diff-15a36057cf9f21154f103a55757a7b59L186), we always called `AdminZkClient#getBrokerMetadatas` regardless of the rack-aware mode. The only difference I can see for the `RackAwareMode.Disabled` case is that the old code would always filter down to the set of brokers register in ZK at the time of the call

```scala
def getBrokerMetadatas(rackAwareMode: RackAwareMode = RackAwareMode.Enforced,
                         brokerList: Option[Seq[Int]] = None): Seq[BrokerMetadata] = {
    val allBrokers = zkClient.getAllBrokersInCluster
    val brokers = brokerList.map(brokerIds => allBrokers.filter(b => brokerIds.contains(b.id))).getOrElse(allBrokers)
    ...
    case RackAwareMode.Disabled => brokers.map(broker => BrokerMetadata(broker.id, None))
```

Does this mean that the new code will generate an assignment for a broker that is currently offline, whereas before it would not?

cmccabe replied:
Good call.  Yeah, we should probably continue the old behavior of ignoring brokers that aren't up at the time of generating the plan, since there's no specific reason to change it.  Fixed.

mumrah reviewed (commented):
Just reviewing the "generate" code path for now. More to follow

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+            } catch {
+              case t: ExecutionException =>
+                t.getCause match {
+                  case _: UnknownTopicOrPartitionException => (topicPartition,
```
The complexity gets a bit deep here. Maybe we could move this try/catch to a separate method

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+   * @param adminClient     The AdminClient to use.
+   */
+  def clearBrokerLevelThrottles(adminClient: Admin): Unit = {
+    val allBrokerIds = adminClient.describeCluster().nodes().get().asScala.map(_.id())
```
Will we miss brokers that are currently offline here?

cmccabe replied:
Yes, we do miss brokers that are currently offline.  This is an existing design flaw with how the reassignment tool works.  I think we can improve it somewhat in this change by at least including the brokers that we know about from the reassignment JSON file.

We can't fix it completely, though, since some brokers might have had quotas set, but not be in the JSON file.  If those brokers are down, they will be missed.  For example, if a broker appeared only as a source broker, it won't be in the JSON file.

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+                                        partitions: Set[TopicPartition])
+                                        : Map[TopicPartition, Seq[Int]] = {
+    val topics = new mutable.HashSet[String]()
+    partitions.foreach(topics += _.topic)
```
Could this just be `partitions.map(_.topic)`?

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+                        brokers: Seq[Int],
+                        enableRackAwareness: Boolean): Seq[BrokerMetadata] = {
+    val brokerSet = brokers.toSet
+    val results = adminClient.describeCluster().nodes().get().asScala.flatMap(
```
nit: using a `filter` followed with a `map` in might be a little more readable here

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+  def verifyBrokerIds(adminClient: Admin, brokers: Set[Int]): Unit = {
+    val allNodeIds = adminClient.describeCluster().nodes().get().asScala.map(_.id).toSet
+    brokers.find(!allNodeIds.contains(_)).map {
+      case id => throw new AdminCommandFailedException(s"Unknown broker id ${id}")
```
Since `id` is not typed, won't this always match? I think the `case` can be removed

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+   */
+  def alterPartitionReassignments(adminClient: Admin,
+                                  reassignments: Map[TopicPartition, Seq[Int]])
+                                
```
Same as question above -- do we need this `case`?

cmccabe replied:
I think to break a tuple apart here, you need a `case`.

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+   */
+  def alterPartitionReassignments(adminClient: Admin,
+                                  reassignments: Map[TopicPartition, Seq[Int]])
+                                
```
Same `case` question

mumrah commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -16,233 +16,1555 @@
+   */
+  def alterPartitionReassignments(adminClient: Admin,
+                                  reassignments: Map[TopicPartition, Seq[Int]])
+                                
```
I'm guessing a `None` is implied here? If so, could we include it?

cmccabe replied:
Github's broken UI is showing this comment as being on line 1170 for me.  Since the comment doesn't make sense for that line, what line did you intend this for?  I checked the file for `case Some` without `case None`, but didn't find anything...

mumrah replied:
After your latest changes, this is L1421 inside `def modifyTopicThrottles`.

```scala
leaderThrottles.get(topicName) match {
  case None =>
  case Some(value) => ...
}
```

cmccabe replied:
OK, then it's already been addressed.  Sounds good.

mumrah reviewed (commented):
In addition to adding support for AdminClient, this PR does a great job refactoring the code to make it more unit testable. Big +1 to that. 

One overall style nit is there is a mixture of `nonEmpty` and `! isEmpty` throughout ReassignPartitionsCommand. I don't personally have a preference, though IntelliJ seems to prefer `nonEmpty`.

Left a few style nits/questions inline

cmccabe commented:
> One overall style nit is there is a mixture of nonEmpty and ! isEmpty throughout ReassignPartitionsCommand. I don't personally have a preference, though IntelliJ seems to prefer nonEmpty
I think part of the confusion is that Java doesn't define `nonEmpty` for collection types, but Scala does.  I'll switch it over to all `nonEmpty` for the Scala collections, though, since that seems more idiomatic for Scala.

mumrah reviewed (approved):
LGTM!
